### PR TITLE
203/account users/container

### PR DESCRIPTION
**Pull request by petermakowski**
## Done

Move filtered users data to Container component for clarity
- do not retry on 404 by default
- handle mutation errors
  - handle additional API errors (e.g. user already belongs to another account)
- better handling of generic fetch network failures
- add missing button types
- colocate user deletion queries with DeleteConfirmationModal

Also:
- add tests for api requests
  - add renderWithQueryClient test util
- add pagination tests
- cleanup AccountUsersData types
- update DeleteConfirmationModal test

## QA
- Go to https://ubuntu-com-10536.demos.haus/advantage/users?test_backend=true and login

## Issue / Card

https://github.com/canonical-web-and-design/commercial-squad/issues/203

## Screenshots

image

### CODE REVIEW

**Comment by wgx** [comment]
@petermakowski Functionally it seems fine, +1

There's some help copy missing from the UI (https://miro.com/app/board/o9J_l9sl6cw=/?moveToWidget=3074457361270970070&cot=14)

On the add user modal we should display this text:

**Billing** users cannot manage users, see tokens or support portal
**Technical** users can access the support portal and see tokens, but not manage users
**Admin** users can perform all actions

I think it's quite a lot to add to the modal, so maybe it could be on an `i` information tooltip next to the role drop-down selector.

**Review by wgx — COMMENTED** [review]
+1 looks great in the UI

**Comment by petermakowski** [comment]
> +1 looks great in the UI

Thanks! Redeploying the demo, the new version will be available soon:

image

**Comment by elioqoshi** [comment]
@petermakowski looks good to me!

Out of curiosity, since I can't reproduce this, what happens when you try to remove yourself (if you are an Admin) while there is another verified admin part of the org?

**Comment by lechsandecki** [comment]
I look at it now and I wonder - is it "add a new user" or "invite a new user" @wgx 
also, I can't see where the "account users" sit in the menu? where would I click to get on that page?
also, is the "account users" the right title - perhaps "Users and roles" would be a better title?
other than that the functionality itself works fine for me ;)
that said, i haven't tested the scenario where i set up a new admin and delete my own account (is it possible if there's a 2nd, validated admin account? can you delete your own account from the user list?)

**Comment by wgx** [comment]
@lechsandecki 
> I look at it now and I wonder - is it "add a new user" or "invite a new user"

It's add - they're not invited, they're instantly added. 

> also, I can't see where the "account users" sit in the menu? where would I click to get on that page?

It's linked from the new advantage ui, but will be in the secondary nav

> also, is the "account users" the right title - perhaps "Users and roles" would be a better title?

I felt like "account users" was more meaningful 

> that said, i haven't tested the scenario where i set up a new admin and delete my own account (is it possible if there's a 2nd, validated admin account? can you delete your own account from the user list?)

It should not be possible to delete your own account - @petermakowski

**Comment by elioqoshi** [comment]
> It should not be possible to delete your own account - @petermakowski

If you cannot delete your own account, the messaging should be different and not `Cannot remove last verified admin from account`

image

**Comment by elioqoshi** [comment]
Eventually it should not even ask you if you are sure whether to delete your own account but just give you an error message right away, no?

**Comment by petermakowski** [comment]
> > It should not be possible to delete your own account - @petermakowski
> 
> If you cannot delete your own account, the messaging should be different and not `Cannot remove last verified admin from account` since it would suggest you can remove your own account if you address this error message.
> 
> image

Could we tweak this on the back-end @albertkol ?

**Review by ClementChaumel — APPROVED** [review]
LGTM

File: static/js/src/advantage/users/AccountUsers.tsx [hunk 1]
```diff
@@ -56,80 +101,54 @@ const AccountUsers = ({
...
   const handleUpdateUser = ({
     email,
     newUserRole,
   }: {
     email: string;
     newUserRole: UserRole;
-  }): Promise<any> =>
+  }) =>
     userUpdateMutation
       .mutateAsync({ email, role: newUserRole })
```
[the remark this answers is not in the document]
Comment (petermakowski) [reply static/js/src/advantage/users/AccountUsers.tsx:124]: I find it useful to be able to chain promises (call me old fashioned).

**Pull request merged**